File: JS/arrayOperations/arrayOperations.js
/**
 * Perform reduce operation over an array
 * 
 * @template A
 * @template T
 * @param reduceStepFunction {(accumulator: A, element: T) => A}
 * @param initialAccumulator {A}
 * @param array {T[]}
 * @returns {A}
 */

function reduce(reduceStepFunction, initialAccumulator, array) {
  TODO()
}

/**
 * Calculate length of an array
 * 
 * @template T
 * @param array {T[]}
 * @returns {number}
 */
function length(array) {
  TODO()
}

/**
 * Return an array with mapFunction applied to all elements of given array
 * 
 * @template T
 * @template S
 * @param mapFunction {(element: T) => S}
 * @param array {T[]}
 * @returns {S[]}
 */
function map(mapFunction, array) {
  TODO()
}

/**
 * Removes elements that do not satisfy the predicate from array
 * 
 * @template T
 * @param predicate {(t: T) => boolean}
 * @array {T[]}
 * @returns {T[]}
 */
function filter(predicate, array) {

}

/**
 * Returns `true` if the array is empty, `false` otherwise
 */

function isEmpty(array) {
  TODO()
}

/**
 * Appends an element to the end of the given array
 */
function append(element, array) {
  TODO()
}

/**
 * Prepends an element to the start of the array
 */
function prepend(element, array) {
  TODO()
}

/**
 * Removes up to `n` elements from the start of the array
 * (returns an empty array if array.length <= n)
 */
function drop(n, array) {
  TODO()
}

/**
 * Takes up to `n` elements from start of the array
 * (returns an array if array.length <= n)
 */
function take(n, array) {
  TODO()
}

function TODO() {
  throw "not implemented"
}

module.exports = {
  append,
  drop,
  filter,
  isEmpty,
  length,
  map,
  prepend,
  reduce
}

File: java/array-operations/src/main/java/com/merck/fun/arrays/ArrayOperations.java
/*
 * Copyright © 2018 Merck Sharp & Dohme Corp., a subsidiary of Merck & Co., Inc.
 * All rights reserved.
 */
package com.merck.fun.arrays;

import java.util.function.Function;
import java.util.function.Predicate;

public class ArrayOperations {

    /**
     * Maps each element of an array to different element
     * @param source - source array to apply map operation to
     * @param mapFunction - function mapping element of type S into T element
     * @param <S> - source type
     * @param <T> - target type
     *
     * @return mapped array
     */
    public static <S, T> T[] map(S[] source, Function<S, T> mapFunction) {
        // TODO:
        return null;
    }

    /**
     * Maps each element of an array to different element
     * @param source - source array to be filtered
     * @param filter - predicate which tells whether to keep or drop given item
     * @param <S> - source type
     *
     * @return filtered array
     */
    public static <S> S[] filter(S[] source, Predicate<S> filter) {
        // TODO:
        return null;
    }

    /**
     * Takes first N elements in the array and returns them as a new array
     * @param source - array to take elements from
     * @param n - number of elements to take
     * @param <S> - source type
     *
     * @return array with first N elements
     */
    public static <S> S[] take(S[] source, int n) {
        // TODO:
        return null;
    }

    /**
     * Joins 2 arrays together
     * @param source - first array to combine
     * @param another - another array to combine
     * @param <S> - source type
     *
     * @return combined array
     */
    public static <S> S[] join(S[] source, S[] another) {
        // TODO:
        return null;
    }

    /**
     * Returns the top N unique elements compared by comparator.
     * @param source - source array
     * @param <S> - source type
     *
     * @return array with top N elements
     */
    public static <S extends Comparable<S>> S[] topN(S[] source, int n) {
        // TODO:
        return null;
    }

    /**
     * Return element which would be ideal candidate as PIVOT for sorting algorithm. Ideal PIVOT item is an element
     * for which half of the elements are greater and half of the elements are lower.
     * @param source - source array
     *
     * @return the ideal PIVOT item
     */
    public static <S extends Comparable<S>> S pivot(S[] source) {
        // TODO:
        return null;
    }
}
